File: 823.带因子的二叉树.c
/*
 * @lc app=leetcode.cn id=823 lang=c
 *
 * [823] 带因子的二叉树
 */

 // @lc code=start
int cmp(const void* p1, const void* p2) {
    return *(int*)p1 - *(int*)p2;
}

int numFactoredBinaryTrees(int* arr, int arrSize) {
    qsort(arr, arrSize,sizeof(int), cmp);
    long long* dp = (long long*)malloc(arrSize * sizeof(long long));
    long long res = 0, mod = 1e9 + 7;
    for (int i = 0; i < arrSize; i++) {
        dp[i] = 1;
        for (int left = 0, right = i - 1;left <= right;left++) {
            while (left <= right && (long long)arr[left] * arr[right] > arr[i]) {
                right--;
            }
            if (left <= right && (long long)arr[left] * arr[right] == arr[i]) {
                if (left == right) {
                    dp[i] = (dp[i] + dp[right] * dp[left]) % mod;
                }
                else {
                    dp[i] = (dp[i] + dp[right] * dp[left] * 2) % mod;
                }
            }
        }
        res = (res + dp[i]) % mod;
    }
    return res;
}
// @lc code=end

File: 823.带因子的二叉树.go
package leetcode

import "sort"

/*
 * @lc app=leetcode.cn id=823 lang=golang
 *
 * [823] 带因子的二叉树
 */

// @lc code=start
func numFactoredBinaryTrees(arr []int) int {
	sort.Ints(arr)
	result, mod := int(0), int(1e9+7)
	dp := make([]int, len(arr))
	for i := 0; i < len(arr); i++ {
		dp[i] = 1
		for left, right := 0, i-1; left <= right; left++ {
			for left <= right && arr[left]*arr[right] > arr[i] {
				right--
			}
			if left <= right && arr[left]*arr[right] == arr[i] {
				if left == right {
					dp[i] = (dp[i] + dp[left]*dp[right]) % mod
				} else {
					dp[i] = (dp[i] + dp[left]*dp[right]*2) % mod
				}
			}
		}
		result = (result + dp[i]) % mod
	}
	return result
}

// @lc code=end
